File: Implementierung/Backend/src/main/java/edu/kit/informatik/model/userdata/rooms/RoomObject.java
package edu.kit.informatik.model.userdata.rooms;

import edu.kit.informatik.model.DataObject;
import edu.kit.informatik.model.User;
import jakarta.persistence.Entity;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.OneToOne;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Cascade;
import org.hibernate.annotations.CascadeType;

import java.sql.Timestamp;

/**
 * Abstrakte Klasse eines Raumobjektes
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Entity
@Getter
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
@NoArgsConstructor
public abstract class RoomObject extends DataObject {

    @OneToOne
    @Cascade(CascadeType.ALL)
    private Position position;

    /**
     * Konstruktor zum Erstellen eines Objektes
     * @param user {@link User}
     * @param position {@link Position}
     * @param createdAt {@link Timestamp} der Erstellung
     * @param updatedAt {@link Timestamp} der letzten Änderung
     */
    public RoomObject(User user, Position position, Timestamp createdAt, Timestamp updatedAt) {
        super(user, createdAt, updatedAt);
        this.position = position;
    }

    /**
     * Rückgabe, ob Raumobjekt ein Tisch ist
     * @return {@code true}, wenn Raumobjekt ein Tisch ist
     */
    public abstract boolean isTable();

    /**
     * Setzen der {@link Position}
     * @param position {@link Position}
     */
    public void setPosition(Position position) {
        this.position = position;
    }
}

File: Implementierung/frontend/paschi/src/model/userdata/rooms/RoomObject.ts
import {Position} from "@/model/userdata/rooms/Position";
import {User} from "@/model/User";
import {Dimensions} from "@/model/userdata/rooms/Dimensions";
import {DataObject} from "@/model/DataObject";

export abstract class RoomObject extends DataObject {

  private readonly _user: User;
  private _position: Position;
  private _dimensions: Dimensions;

  protected constructor(id: string | undefined, localId: number, user: User, position: Position,
                        dimensions: Dimensions) {
    super(id, localId);
    this._user = user;
    this._position = position;
    this._dimensions = dimensions;
  }

  abstract isTable(): boolean;

  get user(): User {
    return this._user;
  }

  get position(): Position {
    return this._position;
  }

  set position(value: Position) {
    this._position = value;
    this.update();
  }

  get dimensions(): Dimensions {
    return this._dimensions;
  }

  set dimensions(value: Dimensions) {
    this._dimensions = value;
    this.update();
  }
}
